sips/sip-clarity4/sip-clarity4.md

```diff
@@ -0,0 +1,466 @@
+    `contract-caller` values to the contract's principal and executes the body
+    expressions within that context, then checks the asset outflows from the
+    contract against the granted allowances, in declaration order. If any
+    allowance is violated, the body expressions are reverted, an error is
+    returned, and an event is emitted with the full details of the violation to
+    help with debugging. Note that the allowance setup expressions are evaluated
+    before executing the body expressions. The final body expression cannot
+    return a `response` value in order to avoid returning a nested `response`
+    value from `as-contract?` (nested responses are error-prone). Returns:
+
+    - `(ok x)` if the outflows are within the allowances, where `x` is the
+      result of the final body expression and has type `A`.
+    - `(err index)` if an allowance was violated, where `index` is the 0-based
+      index of the first violated allowance in the list of granted allowances,
+      or -1 if an asset with no allowance caused the violation.
+
+  - **Example**:
+    ```clarity
+    (define-public (foo)
+      (as-contract? ()
+        (try! (stx-transfer? u1000000 tx-sender recipient))
+      )
+    ) ;; Returns (err -1)
+    (define-public (bar)
+      (as-contract? ((with-stx u1000000))
+        (try! (stx-transfer? u1000000 tx-sender recipient))
+      )
+    ) ;; Returns (ok true)
+    ```
+
+## Built-in keyword: `current-contract`
+
+In many smart contracts, there is a common pattern for retrieving the current
+contract's principal:
+
+```clarity
+(as-contract tx-sender)
+```
+
+With the new `as-contract?` and the removal of `as-contract`, the expression
+required to retrieve the current contract's principal has changed to:
+
+```clarity
+(unwrap-panic (as-contract? () tx-sender))
+```
+
+This new expression is more verbose and less readable. To address this, a new
+built-in keyword `current-contract` is introduced. This keyword provides a
+straightforward way to access the current contract's principal.
+
+- **Description**: Returns the principal of the current contract.
+- **Output**: `principal`
+- **Example**:
+  ```clarity
+  (stx-transfer? u1000000 tx-sender current-contract)
+  ```
+
+## Conversion to `string-ascii`: `to-ascii?`
+
+Originally proposed here: https://github.com/clarity-lang/reference/issues/82
+
+`to-ascii?` is a new Clarity function that converts simple values into their
+`string-ascii` representations.
+
+- **Input**: `int` | `uint` | `bool` | `principal` | `(buff 524284)` |
+  `(string-utf8 262144)`
+- **Output**: `(response (string-ascii 1048571) uint)`
+- **Signature**: `(to-ascii? value)`
+- **Description**: Returns the `string-ascii` representation of the input value
+  in an `ok` response on success. The only error condition is if the input type
+  is `string-utf8` and the value contains non-ASCII characters, in which case,
+  `(err u1)` is returned.
```

Comment:
Do you want a different error for "value is too big to encode?"

Reply:
That's not possible, since it is enforced by the type system.